ff/src/fields/models/fp/montgomery_backend.rs

```diff
@@ -90,9 +97,13 @@ pub trait MontConfig<const N: usize>: 'static + Sync + Send + Sized {
     #[inline(always)]
     fn add_assign(a: &mut Fp<MontBackend<Self, N>, N>, b: &Fp<MontBackend<Self, N>, N>) {
         // This cannot exceed the backing capacity.
-        a.0.add_with_carry(&b.0);
+        let c = a.0.add_with_carry(&b.0);
         // However, it may need to be reduced
-        a.subtract_modulus();
+        if Self::MODULUS_HAS_SPARE_BIT {
+            a.subtract_modulus_with_carry(c);
+        } else {
+            a.subtract_modulus()
+        }
```

Comment:
The logic should be switched, no?
```suggestion
        if Self::MODULUS_HAS_SPARE_BIT {
            a.subtract_modulus()
        } else {
            a.subtract_modulus_with_carry(c)
        }
```